content/blog/2019-11-21-bug-finding/index.md

```diff
@@ -0,0 +1,308 @@
+3:   for (x = 0; x < 5; x++) {
+4:     if (x) continue;
+5:     if (x) break;
+6:   }
+7:   printf("%d", x);
+8: }
+```
+Compiling with LLVM caused this program to print `1` instead of `5`. A loop optimization called "scalar evolution analysis" evaluates loop properties like induction variables and maximum number of iterations. The optimization saw line 5 and incorrectly determined that the loop runs once, meaning `x` must evaluate to `1`.
+
+I find this kind of program interesting because very few programmers in practice would end up writing a function that uses a continue and a break guarded by the same condition. However, when testing optimizations that computes properties of programs, it seems reasonable to test code that never actually runs.
+
+## Design Goals
+The authors provide two design goals for Csmith:
+1. Every randomly generated program must be well formed and have a single interpretation based on the C standard.
+2. Maximize "expressiveness". Expressiveness is the idea that the generated programs should use a wide variety and combinations of language features
+   - For example, Csmith allows for programs with function definitions, global definitions, most C expressions and statements, control flow, arrays, pointers, and more. Some language features that are disallowed include dynamic memory allocation, recursion, and function pointers.
+
+Interestingly, design goal #2 is in direct opposition to design goal #1. The more features we allow to be tested (like arrays and pointers), the easier it is to cause undefined behavior. First, we will discuss at a high level how programs are randomly generated. Then, we will explore how Csmith ensures the generated code doesn't exhibit undefined behavior.
+
+## Randomly Generating Programs
+To generate programs, first Csmith generates types that can be used later during program generation. Then starting from `main`, the authors break down how the program gets filled in.
+1. Csmith selects a random production from the grammar, based on the current context, with potentially non-uniform probability. For example, when inside a function body, Csmith can choose things like `if` and `for` statements, but not a `continue` if the code is not in a loop.
+    - Csmith operates on a subset of C. Interestingly, it doesn't seem like the authors have a grammar explicitly written down for this subset.
+    - For productions that require a *target* (e.g. a variable), Csmith can randomly decide to generate a new target or use an existing one.
+    - Similarly goes for productions requiring a type.
+2. When choosing a nonterminal production, Csmith will recursively choose productions given the appropriate context. For example, if a function call production is chosen, then expressions for the parameters must be generated.
+3. Csmith implements an interprocedural pointer analysis, but needs to keep track of "points-to facts" about the program. As the program is being generated, the set of facts are being updated.
```

Comment:
I'm not sure why the first sentence contains "but." It sounds like those two parts of the sentence say the same thing?

Also, "are being updated" -> "are updated." Or better yet, "Csmith keeps the set of points-to facts up to date while it generates programs." But it would be nice to have a tiny bit more detail about how it manages to maintain alias analysis while doing this seemingly syntactic task.